core/src/View/Kokkos_ViewAlloc.hpp

```diff
@@ -195,26 +195,28 @@ struct ViewValueFunctorSequentialHostInit {
   static_assert(SpaceAccessibility<HostSpace, MemSpace>::accessible);
   static_assert(std::is_same_v<ExecSpace, typename MemSpace::execution_space>);
 
+  ExecSpace space;
   ValueType* ptr;
   size_t n;
+  bool default_exec_space;
 
   ViewValueFunctorSequentialHostInit() = default;
 
   ViewValueFunctorSequentialHostInit(ExecSpace const& arg_space,
                                      ValueType* const arg_ptr,
                                      size_t const arg_n,
                                      std::string /*arg_name*/)
-      : ptr(arg_ptr), n(arg_n) {
-    (void)arg_space;
-    KOKKOS_ASSERT(arg_space == ExecSpace());
-  }
+      : space(arg_space), ptr(arg_ptr), n(arg_n), default_exec_space(false) {}
 
   ViewValueFunctorSequentialHostInit(ValueType* const arg_ptr,
                                      size_t const arg_n,
                                      std::string /*arg_name*/)
-      : ptr(arg_ptr), n(arg_n) {}
+      : space(ExecSpace{}), ptr(arg_ptr), n(arg_n), default_exec_space(true) {}
 
   void construct_shared_allocation() {
+    if (!default_exec_space) {
+      space.fence("Kokkos::View::initialization after allocate");
+    }
```

Comment:
Do we actually need this fence? It's impossible for a provided execution space instance to access the allocation at this point and we guarantee that the allocation mechanism itself always behaves synchronously.

Reply:
It looks like you are right and that CUDA/HIP shared and host pinned allocations are synchronous.
It was unclear to me whether we fence for host accessible SYCL allocations (not saying that is a reason not to drop that fence, just that we might want to check on that).

Reply:
All examples in https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html seem to suggets that `sycl::malloc*` is synchronous: queues are out-of-order by default and there is no fence between kernels and memory allocations (see https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:usm).

Reply:
Yeah, I think we don't need that.